src/file_pipe_log/file_mgr.rs

```diff
@@ -0,0 +1,584 @@
+            handle: fd.handle.clone(),
+            ..*fd
+        })
+    }
+
+    #[inline]
+    pub fn active_file_span(&self) -> (FileSeq, FileSeq) {
+        self.active_files.read().span()
+    }
+
+    #[inline]
+    #[cfg(test)]
+    pub fn stale_file_span(&self) -> (FileSeq, FileSeq) {
+        self.stale_files.read().span()
+    }
+
+    #[inline]
+    pub fn push_back(&mut self, seq: FileSeq, handle: FileWithFormat<F>) -> FileState {
+        self.active_files.write().push_back(seq, handle)
+    }
+
+    /// Rotate a new file handle and return it to the caller.
+    ///
+    /// Returns the relevant `FileSeq`, `PathId` and `F::Handle` of the file
+    /// handle to the caller.
+    /// Attention please, the returned fd should be manually appended to current
+    /// `FileCollection`.
+    pub fn rotate(&mut self) -> Result<(FileSeq, PathId, Arc<F::Handle>)> {
+        let new_file_id = FileId {
+            seq: self.active_files.read().span().1 + 1,
+            queue: self.queue,
+        };
+        let (path_id, new_fd) = if let Some((seq, fd)) = self.stale_files.write().pop_front() {
+            debug_assert_eq!(fd.path_id, 0);
+            let stale_file_id = FileId {
+                seq,
+                queue: self.queue,
+            };
+            let src_path = stale_file_id.build_stale_file_path(&self.paths[fd.path_id]);
+            let dst_path = new_file_id.build_file_path(&self.paths[fd.path_id]);
+            if let Err(e) = self.file_system.reuse(&src_path, &dst_path) {
+                error!("error while trying to reuse one stale file, err: {}", e);
+                if let Err(e) = self.file_system.delete(&src_path) {
+                    error!("error while trying to delete one stale file, err: {}", e);
+                }
+                (fd.path_id, Arc::new(self.file_system.create(&dst_path)?))
+            } else {
+                (fd.path_id, Arc::new(self.file_system.open(&dst_path)?))
+            }
+        } else {
+            let path_id = FileCollection::<F>::get_valid_path(&self.paths, self.target_file_size);
+            let path = new_file_id.build_file_path(&self.paths[path_id]);
+            (path_id, Arc::new(self.file_system.create(&path)?))
+        };
+        Ok((new_file_id.seq, path_id, new_fd))
+    }
+
+    /// Purges files to the specific file_seq.
+    ///
+    /// Returns the purged count of active files.
+    pub fn purge_to(&mut self, file_seq: FileSeq) -> Result<usize> {
+        let (active_state, mut purged_files) = self.active_files.write().split_by(file_seq);
+        if file_seq >= active_state.first_seq + active_state.total_len as FileSeq {
+            debug_assert_eq!(purged_files.len(), 0);
+            return Err(box_err!("Purge active or newer files"));
+        } else if purged_files.len() == 0 {
+            return Ok(0);
+        }
+        let logical_purged_count = purged_files.len();
+        {
+            let mut stale_files = self.stale_files.write();
```

Comment:
```rust
let capacity = self.capacity.saturating_sub(active_state.total_len());
for _ in capacity..stale_files.len() {
  let _ = self.file_system.delete(stale_files.pop_front().unwrap());
}
while stale_files.len() < capacity && purged_files.len() > 0 {
  let f = purged_files.pop_front().unwrap;
  if f.has_log_signing() {
    self.file_system.rename(f);
    stale_files.push_back(f);
  } else {
    let _ = self.file_system.delete(f);
  }
}
```

Reply:
You implementation would not fill the full capacity if there's a V1 purged file. (add a test for this case too).

I updated the previous comment for some edge cases, please just implement this function exactly like it. If there's a difficulty, raise it and let's discuss.